src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -1096,6 +1097,9 @@ std::optional<QueryPipeline> executeInsertSelectWithParallelReplicas(
         pipeline.addCompletedPipeline(std::move(remote_pipeline));
     }
 
+    /// Otherwise CompletedPipelineExecutor uses 1 thread by default
+    pipeline.setNumThreads(settings[Setting::max_threads]);
```

Comment:
Does it make sense to have a runtime check in debug mode if pipeline is being executed in 1 thread despite the `max_threads` settings > 1?

Reply:
I've seen this mistake several times already and we might have even more things like this in the codebase.

Reply:
I remember you did some performance optimizations for S3Queue.